Framework/Geometry/src/Instrument/StructuredDetector.cpp

```diff
@@ -0,0 +1,701 @@
+  shapestr << "<left-back-bottom-point x=\"" << xlb << "\""
+           << " y=\"" << ylb << "\""
+           << " z=\"0\" />";
+  shapestr << "<left-front-bottom-point x=\"" << xlf << "\""
+           << " y=\"" << ylf << "\""
+           << " z=\"0\" />";
+  shapestr << "<right-front-bottom-point x=\"" << xrf << "\""
+           << " y=\"" << yrf << "\""
+           << " z=\"0\" />";
+  shapestr << "<right-back-bottom-point x=\"" << xrb << "\""
+           << " y=\"" << yrb << "\""
+           << " z=\"0\" />";
+  shapestr << "<left-back-top-point x=\"" << xlb << "\""
+           << " y=\"" << ylb << "\""
+           << " z=\"0.001\" />";
+  shapestr << "<left-front-top-point x=\"" << xlf << "\""
+           << " y=\"" << ylf << "\""
+           << " z=\"0.001\" />";
+  shapestr << "<right-front-top-point x=\"" << xrf << "\""
+           << " y=\"" << yrf << "\""
+           << " z=\"0.001\" />";
+  shapestr << "<right-back-top-point x=\"" << xrb << "\""
+           << " y=\"" << yrb << "\""
+           << " z=\"0.001\" />";
+  shapestr << "</hexahedron>";
+  shapestr << "<bounding-box>";
+  shapestr << "<x-min val=\"" << std::min(xlf, xlb) << "\" />";
+  shapestr << "<x-max val=\"" << std::max(xrb, xrf) << "\" />";
+  shapestr << "<y-min val=\"" << ylb << "\" />";
+  shapestr << "<y-max val=\"" << yrf << "\" />";
+  shapestr << "<z-min val=\"0\" />";
+  shapestr << "<z-max val=\"0.001\" />";
+  shapestr << "</bounding-box>";
+  shapestr << "<algebra val=\"" << name << "\" />";
+  shapestr << "</type>";
+
+  boost::shared_ptr<Mantid::Geometry::Object> shape =
+      shapeCreator.createShape(shapestr.str(), false);
+
+  // Create detector
+  auto detector = new Detector(name, id, shape, parent);
+
+  // Set detector position relative to parent
+  V3D pos(xpos, ypos, 0);
+
+  detector->translate(pos);
+
+  return detector;
+}
+
+/** Returns the minimum detector id
+* @return minimum detector id
+*/
+int StructuredDetector::minDetectorID() {
+  if (m_map)
+    return m_base->m_minDetId;
+  return m_minDetId;
+}
+
+//-------------------------------------------------------------------------------------------------
+/** Returns the maximum detector id
+* @return maximum detector id
+*/
+int StructuredDetector::maxDetectorID() {
+  if (m_map)
+    return m_base->m_maxDetId;
+  return m_maxDetId;
+}
+
+boost::shared_ptr<const IComponent>
+StructuredDetector::getComponentByName(const std::string &cname,
+                                       int nlevels) const {
+  // exact matches
+  if (cname == this->getName())
+    return boost::shared_ptr<const IComponent>(this);
+
+  // cache the detector's name as all the other names are longer
+  // The extra ( is because all children of this have that as the next character
+  // and this prevents Bank11 matching Bank 1
+  const std::string MEMBER_NAME = this->getName() + "(";
+
+  // if the component name is too short, just return
+  if (cname.length() <= MEMBER_NAME.length())
```

Comment:
Ok, so this is just an optimization. Why not use `!=` since that would also handle the case that it is too long?